rcl_action/test/rcl_action/test_action_server.cpp

```diff
@@ -0,0 +1,546 @@
+  rcl_reset_error();
+
+  // Get options for an invalid action server
+  rcl_action_server_t invalid_action_server = rcl_action_get_zero_initialized_server();
+  options = rcl_action_server_get_options(&invalid_action_server);
+  EXPECT_EQ(options, nullptr);
+  rcl_reset_error();
+
+  // Get options for a valid action server
+  options = rcl_action_server_get_options(&this->action_server);
+  EXPECT_NE(options, nullptr) << rcl_get_error_string().str;
+}
+
+class TestActionServerCancelPolicy : public TestActionServer
+{
+protected:
+  void SetUp() override
+  {
+    TestActionServer::SetUp();
+    // Add several goals
+    rcl_action_goal_info_t goal_info_in = rcl_action_get_zero_initialized_goal_info();
+    rcl_action_goal_handle_t * goal_handle;
+    rcl_ret_t ret;
+    for (int i = 0; i < NUM_GOALS; ++i) {
+      for (int j = 0; j < UUID_SIZE; ++j) {
+        goal_info_in.uuid[j] = static_cast<uint8_t>(i + j);
+      }
+      goal_handle = rcl_action_accept_new_goal(&this->action_server, &goal_info_in);
+      ASSERT_NE(goal_handle, nullptr) << rcl_get_error_string().str;
+      goal_infos_out[i] = rcl_action_get_zero_initialized_goal_info();
+      ret = rcl_action_goal_handle_get_info(goal_handle, &goal_infos_out[i]);
+      ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+      // Sleep so goals have different acceptance times
+      std::this_thread::sleep_for(std::chrono::milliseconds(250));
+    }
+  }
+
+  void TearDown() override
+  {
+    TestActionServer::TearDown();
+  }
+
+  static const int NUM_GOALS = 10;
+  rcl_action_goal_info_t goal_infos_out[NUM_GOALS];
+};
+
+TEST_F(TestActionServerCancelPolicy, test_action_process_cancel_request_all_goals)
+{
+  // Request to cancel all goals
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  cancel_request.goal_info.stamp.sec = 0;
+  cancel_request.goal_info.stamp.nanosec = 0u;
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    &this->action_server, &cancel_request, &cancel_response);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(cancel_response.msg.goals_canceling.data, nullptr);
+  ASSERT_EQ(cancel_response.msg.goals_canceling.size, (size_t)NUM_GOALS);
+  rcl_action_goal_info_t * goal_info_out;
+  for (int i = 0; i < NUM_GOALS; ++i) {
+    goal_info_out = &cancel_response.msg.goals_canceling.data[i];
+    for (int j = 0; j < UUID_SIZE; ++j) {
+      EXPECT_EQ(goal_info_out->uuid[j], static_cast<uint8_t>(i + j));
+    }
+  }
+}
+
+TEST_F(TestActionServerCancelPolicy, test_action_process_cancel_request_single_goal)
+{
+  // Request to cancel a specific goal
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  for (int i = 0; i < UUID_SIZE; ++i) {
+    cancel_request.goal_info.uuid[i] = static_cast<uint8_t>(i + 2);
+  }
```

Comment:
@jacobperron just curious, why not reusing one of the `init_test_uuid*()` functions?

Reply:
I can reuse `init_test_uui0()`, but the `init_test_uuid1()` wouldn't do because the ID doesn't exist in this test case.

Reply:
Addressed 7aadb55